components/zcash_protocol/src/value.rs

```diff
@@ -0,0 +1,512 @@
+    pub fn from_nonnegative_i64_le_bytes(bytes: [u8; 8]) -> Result<Self, ()> {
+        let amount = i64::from_le_bytes(bytes);
+        Self::from_nonnegative_i64(amount)
+    }
+
+    /// Returns this Zatoshis encoded as a signed two's complement 64-bit
+    /// little-endian value.
+    pub fn to_i64_le_bytes(self) -> [u8; 8] {
+        (self.0 as i64).to_le_bytes()
+    }
+
+    /// Returns whether or not this `Zatoshis` is the zero value.
+    pub fn is_zero(&self) -> bool {
+        self == &Zatoshis::ZERO
+    }
+
+    /// Returns whether or not this `Zatoshis` is positive.
+    pub fn is_positive(&self) -> bool {
+        self > &Zatoshis::ZERO
+    }
+}
+
+impl From<Zatoshis> for ZatBalance {
+    fn from(n: Zatoshis) -> Self {
+        ZatBalance(n.0 as i64)
+    }
+}
+
+impl From<&Zatoshis> for ZatBalance {
+    fn from(n: &Zatoshis) -> Self {
+        ZatBalance(n.0 as i64)
+    }
+}
+
+impl From<Zatoshis> for u64 {
+    fn from(n: Zatoshis) -> Self {
+        n.into_u64()
+    }
+}
+
+impl TryFrom<u64> for Zatoshis {
+    type Error = ();
+
+    fn try_from(value: u64) -> Result<Self, Self::Error> {
+        Zatoshis::from_u64(value)
+    }
+}
+
+impl TryFrom<ZatBalance> for Zatoshis {
+    type Error = ();
+
+    fn try_from(value: ZatBalance) -> Result<Self, Self::Error> {
+        Zatoshis::from_nonnegative_i64(value.0)
+    }
+}
+
+impl Add<Zatoshis> for Zatoshis {
+    type Output = Option<Zatoshis>;
+
+    fn add(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        Self::from_u64(self.0.checked_add(rhs.0)?).ok()
+    }
+}
+
+impl Add<Zatoshis> for Option<Zatoshis> {
+    type Output = Self;
+
+    fn add(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        self.and_then(|lhs| lhs + rhs)
+    }
+}
+
+impl Sub<Zatoshis> for Zatoshis {
+    type Output = Option<Zatoshis>;
+
+    fn sub(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        Zatoshis::from_u64(self.0.checked_sub(rhs.0)?).ok()
+    }
+}
+
+impl Sub<Zatoshis> for Option<Zatoshis> {
+    type Output = Self;
+
+    fn sub(self, rhs: Zatoshis) -> Option<Zatoshis> {
+        self.and_then(|lhs| lhs - rhs)
+    }
+}
+
+impl Mul<usize> for Zatoshis {
+    type Output = Option<Self>;
+
+    fn mul(self, rhs: usize) -> Option<Zatoshis> {
+        Zatoshis::from_u64(self.0.checked_mul(u64::try_from(rhs).ok()?)?).ok()
+    }
+}
+
+impl Sum<Zatoshis> for Option<Zatoshis> {
+    fn sum<I: Iterator<Item = Zatoshis>>(iter: I) -> Self {
+        iter.fold(Some(Zatoshis::ZERO), |acc, a| acc? + a)
+    }
+}
+
+impl<'a> Sum<&'a Zatoshis> for Option<Zatoshis> {
+    fn sum<I: Iterator<Item = &'a Zatoshis>>(iter: I) -> Self {
+        iter.fold(Some(Zatoshis::ZERO), |acc, a| acc? + *a)
+    }
+}
+
+/// A type for balance violations in amount addition and subtraction
+/// (overflow and underflow of allowed ranges)
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum BalanceError {
+    Overflow,
+    Underflow,
+}
```

Comment:
It's somewhat confusing that we define `BalanceError` here but we don't use it in this module (instead clients end up having to do `.ok_or(BalanceError::Underflow)` or `.ok_or(BalanceError::Overflow)`).